src/shared/index.ts

```diff
@@ -0,0 +1,211 @@
+				return [ value.line, value.column, valueEnd.line, valueEnd.column ] as _Region
+			})()
+
+			const ploc = result.locations?.[0]?.physicalLocation
+			const [uri, uriContents] = parseArtifactLocation(result, ploc?.artifactLocation)
+			result._uri = uri
+			result._uriContents = uriContents
+			{
+				const parts = uri?.split('/')
+				implicitBase = // Base calc (inclusive of dash for now)
+					implicitBase?.slice(0, Array.commonLength(implicitBase, parts ?? []))
+					?? parts
+				const file = parts?.pop()
+				if (file && uri) {
+					fileAndUris.push([file, uri.replace(/^\//, '')]) // Normalize leading slashes.
+				}
+			}
+			result._region = parseRegion(ploc?.region)
+			result._line = result._region?.[0] ?? result._region ?? -1 // _line is sugar for _region
+
+			result._rule = run.tool.driver.rules?.[result.ruleIndex] // If result.ruleIndex is undefined, that's okay.
+			const message = result._rule?.messageStrings?.[result.message.id] ?? result.message
+			result._message = format(message.text || result.message?.text, result.message.arguments) ?? '—'
+			result._markdown = format(message.markdown || result.message?.markdown, result.message.arguments) // No '—', leave undefined if empty.
+
+			result.level = result.level ?? 'warning'
+			result.baselineState = result.baselineState ?? 'new'
+			result._suppression = !result.suppressions || result.suppressions.every(sup => sup.status === 'rejected')
+				? 'not suppressed'
+				: 'suppressed'
+		})
+
+		run._implicitBase = implicitBase?.join('/')
+		run.results?.forEach(result => {
+			result._relativeUri = result._uri?.replace(run._implicitBase , '') ?? '' // For grouping, Empty works more predictably than undefined
+		})
+	})
+	log._distinct = mapDistinct(fileAndUris)
+	log._jsonMap = undefined // Free-up memory.
+}
+
+/*
+TfLoc
+   location: Loc
+
+Result
+   locations: Loc[]
+
+Loc
+   Message
+   PhyLoc
+      ArtLoc: Uri, Index
+      Region
+
+Run.artifacts: Art[]
+   location: ArtLoc
+   contents: ArtCon
+*/
+export function parseLocation(result: Result, loc?: Location) {
+	const message = loc?.message?.text
+	const [uri, uriContent] = parseArtifactLocation(result, loc?.physicalLocation?.artifactLocation)
+	const region = loc?.physicalLocation?.region
+	return { message, uri, uriContent, region }
+}
+
+export function parseRegion(region: Region): _Region {
+	if (!region) return undefined
+
+	const {byteOffset, byteLength} = region
+	if (byteOffset !== undefined && byteLength !== undefined) return [byteOffset, byteLength] as [number, number]
+
+	let {startLine, startColumn, endLine, endColumn} = region
+	if (!startLine) return undefined // Lines are 1-based so no need to check undef.
+
+	startLine--
+	if (!startColumn) return startLine
+
+	startColumn--
+	if (endColumn) endColumn--
+	if (endLine) endLine--
+	return [
+		startLine,
+		startColumn,
+		endLine ?? startLine,
+		endColumn ?? (startColumn + 1)
+	] as [number, number, number, number]
```

Comment:
> as [number, number, number, number] [](start = 2, length = 36)

Isn't this redundant?

Reply:
Typescript infers this as `number[]`. We want a tuple of exact four numbers.

Reply:
perhaps make an interface for this

---
In reply to: [431331282](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r431331282) [](ancestors = 431331282)

Reply:
Will do.

---
In reply to: [434794240](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r434794240) [](ancestors = 434794240,431331282)

Reply:
Typescript infers this as number[]. We want a tuple of exactly four numbers.

---
In reply to: [431219212](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r431219212) [](ancestors = 431219212)

Reply:
Added more explicit types for this.

---
In reply to: [436197296](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r436197296) [](ancestors = 436197296,434794240,431331282)